docs/contributing/provider-design.md

```diff
@@ -81,6 +84,20 @@ The rationale behind this design decision includes the following:
 
 A poignant real-world example of the last point involved a Lambda resource. The resource helped clean up extra resources (ENIs) due to a common misconfiguration. Practitioners found the functionality helpful since the issue was hard to diagnose. Years later, AWS updated the Lambda API. Immediately, practitioners reported that Terraform executions were failing. Downgrading the provider was not possible since many configurations depended on recent releases. For environments running many versions behind, forcing an upgrade with the fix would likely cause unrelated and unexpected changes. In the end, HashiCorp and AWS performed a large-scale outreach to help upgrade and fixing the misconfigurations. Provider maintainers and practitioners lost considerable time.
 
+### Data Sources
+
+A separate class of Terraform resource types are [data sources](https://www.terraform.io/docs/language/data-sources/). These are typically intended as a configuration method to lookup or fetch data in a read-only manner. Data sources should not have side effects on the remote system.
+
+When discussing data sources, they are typically classified by the intended number of return objects or data. Singular data sources represent a one-to-one lookup or data operation. Plural data sources represent a one-to-many operation.
+
+#### Plural Data Sources
+
+These data sources are intended to return zero, one, or many results, usually associated with a managed resource type. Typically results are a set unless ordering guarantees are provided by the remote system. These should be named with a plural suffix (e.g. `s` or `es`) and should not include any specific attribute in the naming (e.g. prefer `aws_ec2_transit_gateways` instead of `aws_ec2_transit_gateway_ids`).
```

Comment:
We have not been consistent about handling the case when a query-for-multiple returns 0 results.
There are a number of open issues around this, most generally #17540 and #8302.

Reply:
Yep, this documentation serves to move towards standardizing data source behaviors in a consistent manner across the provider. Eventually we can (and should) go back through any implementations not following this guidance and remove errors for zero results.

Previously, the Terraform CLI maintainers were hesitant with zero result data sources due to their graph placement and the potential for unexpected behaviors, but since Terraform CLI 0.14 includes these nodes in the graph just like managed resources, it is a little safer to rely on the attribute values further down the graph.